compiler/rustc_hir_typeck/src/method/probe.rs

```diff
@@ -1218,8 +1218,54 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
         )>,
         mut unstable_candidates: Option<&mut Vec<(Candidate<'tcx>, Symbol)>>,
     ) -> Option<PickResult<'tcx>> {
+        // Delegation item can be expanded into method calls or fully qualified calls
+        // depending on the callee's signature. Method calls are used to allow
+        // autoref/autoderef for target expression. For example in:
+        //
+        // trait Trait : Sized {
+        //     fn by_value(self) -> i32 { 1 }
+        //     fn by_mut_ref(&mut self) -> i32 { 2 }
+        //     fn by_ref(&self) -> i32 { 3 }
+        // }
+        //
+        // struct NewType(SomeType);
+        // impl Trait for NewType {
+        //     reuse Trait::* { self.0 }
+        // }
+        //
+        // `self.0` will automatically coerce. The difference with existing method lookup
+        // is that methods in delegation items are pre-resolved by callee path (`Trait::*`).
+        // Therefore, we are only forced to consider the pre-resolved method here.
+        let delegation_res = if self.fcx.tcx.hir().opt_delegation_sig_id(self.body_id).is_some() {
+            let body = self.tcx.hir().body_owned_by(self.body_id);
+            let block = match body.value.kind {
+                hir::ExprKind::Block(block, _) => block,
+                _ => unreachable!(),
+            };
+            let expr = block.expr.unwrap();
+            let res_id = match expr.kind {
+                hir::ExprKind::MethodCall(segment, ..) => segment.res.def_id(),
+                hir::ExprKind::Call(
+                    hir::Expr { kind: hir::ExprKind::Path(hir::QPath::Resolved(_, path)), .. },
+                    _,
+                ) => path.res.def_id(),
+                _ => unreachable!(),
+            };
+            Some((expr.hir_id, res_id))
+        } else {
+            None
+        };
         let mut applicable_candidates: Vec<_> = candidates
             .iter()
+            .filter(|candidate| {
+                if let Some((expr_id, res_id)) = delegation_res
+                    && self.scope_expr_id == expr_id
+                    && candidate.item.def_id != res_id
```

Comment:
Expected resolution already lives in `PathSegment` of the resolved method call (any `PathSegment::res` that is not `Res::Err` is the expected resolution).
I think its better to pass it from places like `resolve_fully_qualified_call` to this in the same exact way as `scope_expr_id`.

If that is too invasive then it's better to make a separate table `Map<HirId, Res>` similar to `trait_map` and lookup `scope_expr_id` in it.
Doing HIR inspection like in the code above is probably the worst alternative.